run_matrix.py

```diff
@@ -251,69 +281,164 @@ def _test_roles(
         )
 
     return results_per_version
 
 
 def run_interface_tests(
-    path: Path, include: str = "*", keep_cache: bool = False
+    path: Path,
+    repo: str,
+    branch: str,
+    include: str = "*",
+    keep_cache: bool = False,
 ) -> "_ResultsPerInterface":
     """Run the tests for the specified interfaces, defaulting to all."""
     if not keep_cache:
         _clean()
     test_results = {}
     collected = collect_tests(path=path, include=include)
     for interface, version_to_roles in collected.items():
-        results_per_version = _test_interface_version(version_to_roles, interface)
+        results_per_version = _test_interface_version(
+            version_to_roles, interface, repo, branch
+        )
         test_results[interface] = results_per_version
 
+        # running in github actions with owners set on the test
+        if os.getenv("GITHUB_ACTIONS"):
+            for version, tests_per_role in version_to_roles.items():
+                owners = tests_per_role.get("owners")
+                if owners and test_failed(results_per_version[version]):
+                    create_issue(
+                        interface, version, results_per_version[version], owners
+                    )
+
     if not collected:
         logging.warning("No tests collected.")
 
     return test_results
 
 
+def test_failed(role_result: "_ResultsPerRole"):
+    for _, charm_result in role_result.items():
+        if False in charm_result.values():
+            return True
+
+    return False
+
+
+def create_issue(
+    interface: str, version: str, result_per_role: "_ResultsPerRole", owners: List[str]
+):
+    github_token = os.getenv("GITHUB_TOKEN")
+    g = Github(github_token)
+    repo = g.get_repo("canonical/charm-relation-interfaces")
+    workflow_url = ""
+    github_run_id = os.getenv("GITHUB_RUN_ID")
+    if github_run_id:
+        workflow_url = f"https://github.com/canonical/charm-relation-interfaces/actions/runs/{github_run_id}"
+    result = flatten_test_result(result_per_role)
+    title = f"Interface test for {interface} {version} failed."
+    body = f"""\
+Tests for interface {interface} {version} failed.
+
+{result}
+
+See the workflow {workflow_url} for more detail.
+"""
+
+    issue = None
+    for i in repo.get_issues(state="open"):
+        if f"{interface} {version}" in i.title:
+            issue = i
+            break
+
+    if not issue:
+        issue = repo.create_issue(title=title, body=body)
+        print(f"GitHub issue created: {issue.html_url}")
+    else:
+        issue.create_comment(body)
+        print(f"GitHub issue updated: {issue.html_url}")
+
+    if owners:
+        issue.edit(assignees=owners)
+        print(f"GitHub issue assigned to {owners}")
+
+
+def flatten_test_result(version_result: "_ResultsPerRole"):
+    result = ""
+
+    provider_res = ""
+    for charm, res in version_result.get("provider").items():
+        if res is False:
+            provider_res += f"- {charm}: {res}\n"
```

Comment:
My inclination is that this (and the one below) would be better like:

```suggestion
    provider_res = "\n".join(
        f"- {charm}: {res}"
        for charm, res in version_result.get("provider").items()
        if res is False
    )
```

However, maybe that's just because I spent so much time avoiding string concatenation in the days that was expensive in CPython. If you prefer it as you have it, that's fine.

Reply:
It looks nicer, plus it avoids string concatenation, I will change it.